src/backend/optimizer/util/plancat.c

```diff
@@ -546,12 +587,31 @@ cdb_estimate_rel_size(RelOptInfo   *relOptInfo,
 		int32		tuple_width;
 
 		tuple_width = get_rel_data_width(rel, attr_widths);
-		tuple_width += sizeof(HeapTupleHeaderData);
-		tuple_width += sizeof(ItemPointerData);
+
+		if (RelationIsAoRows(rel))
+		{
+			/*
+			 * Row oriented relations store memtuples, that have a header,
+			 * plus storage is done in VarBlock, which has item-offsets array
+			 * with at 2 or 3 byte offsets. For simplicity, in our rough
+			 * estimation we consider only 2 byte offsets. Thus, we add
+			 * (2 + header size) to the tuple width.
+			 */
+			tuple_width += (offsetof(MemTupleData, PRIVATE_mt_bits) + 2);
+		}
+		else if (!RelationIsAoCols(rel))
+		{
+			/* do the adjustment for heap relations */
+			tuple_width += sizeof(HeapTupleHeaderData);
+			tuple_width += sizeof(ItemPointerData);
+		}
+
 		/* note: integer division is intentional here */
 		density = (BLCKSZ - SizeOfPageHeaderData) / tuple_width;
 	}
-	*tuples = rint(density * (double) curpages);
+	/* Calculate tuples only if they were not set before */
+	if (*tuples < 0)
```

Comment:
Can we place `density` calculation under this condition? `density` is used to get `tuples` only